metaflow/plugins/datatools/s3/s3op.py

```diff
@@ -322,22 +328,60 @@ def op_info(url):
                             # We indicate that the file was uploaded
                             result_file.write("%d %d\n" % (idx, 0))
                         except client_error as err:
-                            error_code = normalize_client_error(err)
-                            if error_code == 403:
-                                result_file.write(
-                                    "%d %d\n" % (idx, -ERROR_URL_ACCESS_DENIED)
-                                )
-                                continue
-                            elif error_code == 503:
-                                result_file.write("%d %d\n" % (idx, -ERROR_TRANSIENT))
-                                continue
-                            else:
-                                raise
+                            # shouldn't get here, but just in case
+                            handle_client_error(err, idx, result_file)
+                            continue
+                        except S3UploadFailedError as e:
+                            err = convert_to_client_error(e)
+                            handle_client_error(err, idx, result_file)
+                            continue
+                        except (SSLError, Exception) as e:
+                            # assume anything else is transient
+                            result_file.write("%d %d\n" % (idx, -ERROR_TRANSIENT))
+                            result_file.flush()
+                            continue
         except:
             traceback.print_exc()
+            result_file.flush()
             sys.exit(ERROR_WORKER_EXCEPTION)
 
 
+def convert_to_client_error(e):
+    match = re.search(
+        r"An error occurred \((\w+)\) when calling the (\w+) operation.*: (.+)", str(e)
```

Comment:
nit: pre-compile this expression

Reply:
Do we need to do this matching? The error format is not an interface guaranteed by boto3 and can change at any point.

Reply:
I don't see how we can tell if it's a fatal error or a transient error without doing the matching. Do you have some ideas?